### DISCUSSION TITLE
Integrated Git: Permission denied (publickey)

### DISCUSSION THREAD

**Discussion by Lestt** [coder/code-server#3641] [General]
Hi all,

I'm using code-server on my ChromeBook. So it is running inside a Debiant 10 buster container.
In the Debian, the code-server is launched as a service, when the container is started.

When i'm using the integrated Git feature in code-server, i am not able to pull / push my git project because i need to authenticate via ssh key. Here the error in the git output:

```
Looking for git in: git
Using git 2.20.1 from git
> git rev-parse --git-dir
Open repository: /home/lest/my_project
> git fetch
> git status -z -u
> git symbolic-ref --short HEAD
> git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track) refs/heads/my-branch refs/remotes/my-branch
> git for-each-ref --sort -committerdate --format %(refname) %(objectname) %(*objectname)
> git remote --verbose
> git config --get commit.template
git@bitbucket.org: Permission denied (publickey).
fatal: Could not read from remote repository.
```

When i'm trying to use git via the integrated teminal, it is working well, after asking for my passphrase. And it is the same behavior as in my regular terminal.

Obviously the integrated terminal inherit my shell env, that's why it is working well. But the integrated Git interface is not able to grab the ssh configuration.

In my ~/.ssh/config, I've added:
```
Host bitbucket.org
     HostName bitbucket.org
     AddKeysToAgent yes 
     User <lest@lest.com>
     IdentityFile ~/.ssh/lest_id_rsa
     IdentitiesOnly yes                 
```

My ssh-agent is launched as a service as well, and it is running:
```
● ssh-agent.service - SSH key agent
   Loaded: loaded (/usr/lib/systemd/system/ssh-agent.service; enabled; vendor preset: enabled)
   Active: active (running) since ...
 Main PID: 91 (ssh-agent)
    Tasks: 1 (limit: 4915)
   Memory: 1.0M
   CGroup: /system.slice/ssh-agent.service
           └─91 /usr/bin/ssh-agent -D -a /run/ssh-agent.socket
```

And i have this file (not sure if it useful):
```
$ cat ~/.pam_environment
SSH_AUTH_SOCK DEFAULT="${XDG_RUNTIME_DIR}/ssh-agent.socket"
```

I'm really running out of ideas why I can't make it work. Some help will be really appreciated !
Thanks.

**Comment by jsjoeio** [comment]
Hmm...I feel like @code-asher has helped someone with this in the past 🤔 Maybe he has some insight.

**Comment by code-asher** [comment]
I suppose the UI is unable to ask for key's passphrase. I know VS Code
has some measures in place to make user/password authentication work but
I'm not sure it has anything to make the agent work.

Local VS Code doesn't have to worry about this because the pinentry
prompt will pop up normally but with the web something else needs to be
done.

Related:

https://github.com/cdr/code-server/issues/3189#issuecomment-826951098
https://github.com/cdr/code-server/issues/695#issuecomment-589784243

**Reply by code-asher** [reply]
The only workaround I can think of at the moment is to unlock the key on the terminal before using the UI.

**Reply by Lestt** [reply]
> The only workaround I can think of at the moment is to unlock the key on the terminal before using the UI.

@code-asher It is working well if you are running `code-server` via your shell. But in my case it is a service launched at boot. And I can't figure it out is there is a way to achieve the same with the service. Maybe a conf in the service file ? I didn't find nothing on Google about it.

I have already seen the two related links you have provided here. Indeed, I think the system is trying to authenticate, and a popup windows should appears and ask for my password. I don't know if it's feasable here.

**Reply by code-asher** [reply]
Unfortunately I don't have any experience with ssh-agent so I'm not sure how to work around it. With gpg-agent I either use `--pinentry-mode=loopback` or set the pinentry to use `pinentry-tty` so maybe ssh-agent has similar things.

**Reply by vegardhw** [reply]
Same thing happens with the VSCode native app and remote development with Remote SSH to another server which is using passphrase protected SSH keys for repos. Git push in VSCode GUI will not work but terminal is OK.
